In [1]:
import graphlab as gl

In [2]:
# read product review data
# for baby products
products = gl.SFrame('../datasets/amazon_baby.gl/')

This non-commercial license of GraphLab Create for academic use is assigned to pialgi@live.com and will expire on March 28, 2018.


[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: /tmp/graphlab_server_1492993760.log


In [3]:
#set the output target to be this window
gl.canvas.set_target('ipynb')

In [4]:
# now we see that there are 68% of 5 star reviews

In [5]:
#it is time to build a sentiment classifier

In [6]:
#we'll ignore all reviews with rating=3, since 
#they are neutral reviews. So everything over 3 is considered a positive 
#less then 3 a negative

In [7]:
#filter out 3* reviews
products=products[products['rating'] != 3]

In [8]:
#now we are going to use boolean values to quantify sentiment review
# positive = 1, negative = 0
# extra column will be created to hold these values
products['sentiment'] = products['rating'] >= 4

In [9]:
selected_words = ['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']

In [10]:
products['word_count'] = gl.text_analytics.count_words(products['review'])

In [11]:
print(selected_words)

['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']


In [12]:
for word in selected_words :
    f_name = word
    def f_name(d) :
         if word in d :
            return d[word]
         else :
            return 0
    
    products[word] = products['word_count'].apply(f_name)

In [13]:
products.head()

Columns:
	name	str
	review	str
	rating	float
	sentiment	int
	word_count	dict
	awesome	int
	great	int
	fantastic	int
	amazing	int
	love	int
	horrible	int
	bad	int
	terrible	int
	awful	int
	wow	int
	hate	int

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+-----------+
|              name             |             review            | rating | sentiment |
+-------------------------------+-------------------------------+--------+-----------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |     1     |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |     1     |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |     1     |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |     1     |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |     1     |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |     1     |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |     1     |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |     1     |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |     1     |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |     1     |
+-------------------------------+-------------------------------+--------+-----------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'and': 3, 'love': 1, 'it'... |    0    |   0   |     0     |    0    |  1   |
| {'and': 2, 'quilt': 1, 'it... |    0    |   0   |     0     |    0    |  0   |
| {'ingenious': 1, 'and': 3,... |    0    |   0   |     0     |    0    |  2   |
| {'and': 2, 'parents!!': 1,... |    0    |   1   |     0     |    0    |  0   |
| {'and': 2, 'this': 2, 'her... |    0    |   1   |     0     |    0    |  0   |
| {'shop': 1, 'noble': 1, 'i... |    0    |   0   |     0     |    0    |  0   |
| {'and': 2, 'all': 1, 'righ... |    0    |   0   |     0     |    0    |  0   |
| {'and': 1, 'help': 1, 'giv... |    0    |   0   |     0     |    0    |  0   |
| {'journal.': 1, 'nanny': 1... |    0    |   0   |     0     |    0    |  0   |
| {'all': 1, 'forget': 1, 'j... |    0    |   0   |     0     |    0    |  2   |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+
| horrible | bad | terrible | awful | wow | hate |
+----------+-----+----------+-------+-----+------+
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
|    0     |  0  |    0     |   0   |  0  |  0   |
+----------+-----+----------+-------+-----+------+
[10 rows x 16 columns]

In [14]:
#for word in selected_words :
#    del products[word]

In [15]:
max = 0
best_word = 'baaa'
for word in selected_words:
    ss = products[word].sum()
    if ss > max :
        max = ss
        best_word = word
        

In [16]:
min = 999999999
worst_word = 'baaa'
for word in selected_words:
    mm = products[word].sum()
    if mm < min :
        min = mm
        worst_word = word

In [17]:
print max , best_word, min, worst_word

42420 great 131 wow


# THE BEST WORD IS GREAT 42420

In [18]:
products.column_names()

['name',
 'review',
 'rating',
 'sentiment',
 'word_count',
 'awesome',
 'great',
 'fantastic',
 'amazing',
 'love',
 'horrible',
 'bad',
 'terrible',
 'awful',
 'wow',
 'hate']

In [19]:
train_data,test_data = products.random_split(.8, seed= 0)

In [20]:
selected_words_model = gl.logistic_classifier.create(
    train_data,target='sentiment',features=selected_words,
          validation_set=test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 11

Number of unpacked features : 11

Number of coefficients    : 12

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Elapsed Time | Training-accuracy | Validation-accuracy |

+-----------+----------+--------------+-------------------+---------------------+

| 1         | 2        | 1.299829     | 0.844299          | 0.842842            |

| 2         | 3        | 1.485697     | 0.844186          | 0.842842            |

| 3         | 4        | 1.671131     | 0.844276          | 0.843142            |

| 4         | 5        | 1.859274     | 0.844269          | 0.843142            |

| 5         | 6        | 2.056609     | 0.844269          | 0.843142            |

| 6         | 7        | 2.255049     | 0.844269          | 0.843142            |

+-----------+----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [21]:
selected_words_model['coefficients'].head()

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 10

Data:
+-------------+-------+-------+-----------------+------------------+
|     name    | index | class |      value      |      stderr      |
+-------------+-------+-------+-----------------+------------------+
| (intercept) |  None |   1   |  1.36728315229  | 0.00861805467824 |
|   awesome   |  None |   1   |  1.05800888878  |  0.110865296265  |
|    great    |  None |   1   |  0.883937894898 | 0.0217379527921  |
|  fantastic  |  None |   1   |  0.891303090304 |  0.154532343591  |
|   amazing   |  None |   1   |  0.892802422508 |  0.127989503231  |
|     love    |  None |   1   |  1.39989834302  | 0.0287147460124  |
|   horrible  |  None |   1   |  -1.99651800559 | 0.0973584169028  |
|     bad     |  None |   1   | -0.985827369929 | 0.0433603009142  |
|   terrible  |  None |   1   |  -2.09049998487 | 0.0967241912229  |
|    awful    |  None |   1   |  -1.76469955631 |  0.134679803365  |
+-------------+-------+-------+-----------------+------------------+
[10 rows x 5 columns]

In [22]:
selected_words_model['coefficients'].sort('value')

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 12

Data:
+-----------+-------+-------+------------------+-----------------+
|    name   | index | class |      value       |      stderr     |
+-----------+-------+-------+------------------+-----------------+
|  terrible |  None |   1   |  -2.09049998487  | 0.0967241912229 |
|  horrible |  None |   1   |  -1.99651800559  | 0.0973584169028 |
|   awful   |  None |   1   |  -1.76469955631  |  0.134679803365 |
|    hate   |  None |   1   |  -1.40916406276  | 0.0771983993506 |
|    bad    |  None |   1   | -0.985827369929  | 0.0433603009142 |
|    wow    |  None |   1   | -0.0541450123333 |  0.275616449416 |
|   great   |  None |   1   |  0.883937894898  | 0.0217379527921 |
| fantastic |  None |   1   |  0.891303090304  |  0.154532343591 |
|  amazing  |  None |   1   |  0.892802422508  |  0.127989503231 |
|  awesome  |  None |   1   |  1.05800888878   |  0.110865296265 |
+-----------+-------+-------+------------------+-----------------+
[12 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [23]:
selected_words_model['coefficients'].sort('value').tail()

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 10

Data:
+-------------+-------+-------+------------------+------------------+
|     name    | index | class |      value       |      stderr      |
+-------------+-------+-------+------------------+------------------+
|    awful    |  None |   1   |  -1.76469955631  |  0.134679803365  |
|     hate    |  None |   1   |  -1.40916406276  | 0.0771983993506  |
|     bad     |  None |   1   | -0.985827369929  | 0.0433603009142  |
|     wow     |  None |   1   | -0.0541450123333 |  0.275616449416  |
|    great    |  None |   1   |  0.883937894898  | 0.0217379527921  |
|  fantastic  |  None |   1   |  0.891303090304  |  0.154532343591  |
|   amazing   |  None |   1   |  0.892802422508  |  0.127989503231  |
|   awesome   |  None |   1   |  1.05800888878   |  0.110865296265  |
| (intercept) |  None |   1   |  1.36728315229   | 0.00861805467824 |
|     love    |  None |   1   |  1.39989834302   | 0.0287147460124  |
+-------------+-------+-------+------------------+------------------+
[10 rows x 5 columns]

# THE MOST NEGATIVE WEIGHT IS TERRIBLE -2.09

# the most positive 1.39989834302

In [24]:
selected_words_model.evaluate(test_data)

{'accuracy': 0.8431419649291376,
 'auc': 0.6648096413721418,
 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      0       |        0        |  234  |
 |      0       |        1        |  5094 |
 |      1       |        1        | 27846 |
 |      1       |        0        |  130  |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.914242563530107,
 'log_loss': 0.4054747110365649,
 'precision': 0.8453551912568306,
 'recall': 0.9953531598513011,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+-----+-----+-------+------+
 | threshold | fpr | tpr |   p   |  n   |
 +-----------+-----+-----+-------+------+
 |    0.0    | 1.0 | 1.0 | 27976 | 5328 |
 |   1e-05   | 1.0 | 1.0 | 27976 | 5328 |
 |   2e-05   | 

# SELECTED WORDS ACCURACY 0.84
# sentiment accuracy 0.91


In [25]:
selected_words_model.show(view='Evaluation')

In [26]:
diaper_champ_reviews = products[products['name']=='Baby Trend Diaper Champ']

In [27]:
len(diaper_champ_reviews)

298

In [28]:
diaper_champ_reviews['predicted_sentiment'] = selected_words_model.predict(diaper_champ_reviews,output_type='probability')

In [29]:
diaper_champ_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	sentiment	int
	word_count	dict
	awesome	int
	great	int
	fantastic	int
	amazing	int
	love	int
	horrible	int
	bad	int
	terrible	int
	awful	int
	wow	int
	hate	int
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------+-------------------------------+--------+-----------+
|           name          |             review            | rating | sentiment |
+-------------------------+-------------------------------+--------+-----------+
| Baby Trend Diaper Champ | Ok - newsflash.  Diapers a... |  4.0   |     1     |
| Baby Trend Diaper Champ | My husband and I selected ... |  1.0   |     0     |
| Baby Trend Diaper Champ | Excellent diaper disposal ... |  5.0   |     1     |
| Baby Trend Diaper Champ | We love our diaper champ. ... |  5.0   |     1     |
| Baby Trend Diaper Champ | Two girlfriends and two fa... |  5.0   |     1     |
| Baby Trend Diaper Champ | I waited to review this un... |  4.0   |     1     |
| Baby Trend Diaper Champ | I have had a diaper genie ... |  1.0   |     0     |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |     1     |
| Baby Trend Diaper Champ | I am so glad I got the Dia... |  5.0   |     1     |
| Baby Trend Diaper Champ | We had 2 diaper Genie's bo... |  4.0   |     1     |
+-------------------------+-------------------------------+--------+-----------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'son': 1, 'just': 2, 'les... |    0    |   0   |     0     |    0    |  0   |
| {'material)': 1, 'bags,': ... |    0    |   0   |     0     |    0    |  0   |
| {'control': 1, 'am': 1, 'i... |    0    |   0   |     0     |    0    |  0   |
| {'and': 3, 'over.': 1, 'al... |    0    |   0   |     0     |    0    |  1   |
| {'just': 1, '-': 3, 'both'... |    0    |   0   |     0     |    0    |  0   |
| {'lysol': 1, 'all': 1, 'mo... |    0    |   0   |     0     |    0    |  0   |
| {'all': 1, 'bags.': 1, 'ju... |    0    |   0   |     0     |    0    |  0   |
| {'lysol': 1, 'all': 2, 'ba... |    0    |   0   |     0     |    0    |  0   |
| {'and': 2, 'all': 1, 'just... |    0    |   0   |     0     |    0    |  0   |
| {'hand.': 1, 'both': 1, '(... |    0    |   0   |     0     |    0    |  2   |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+---------------------+
| horrible | bad | terrible | awful | wow | hate | predicted_sentiment |
+----------+-----+----------+-------+-----+------+---------------------+
|    0     |  0  |    0     |   0   |  0  |  0   |    0.796940851291   |
|    0     |  0  |    0     |   0   |  0  |  0   |    0.796940851291   |
|    0     |  0  |    0     |   0   |  0  |  0   |    0.796940851291   |
|    0     |  0  |    0     |   0   |  0  |  0   |    0.940876393428   |
|    1     |  0  |    0     |   0   |  0  |  0   |    0.347684052736   |
|    0     |  1  |    0     |   0   |  0  |  0   |     0.5942241719    |
|    0     |  0  |    0     |   0   |  0  |  0   |    0.796940851291   |
|    0     |  0  |    0     |   0   |  0  |  0   |    0.796940851291   |
|    0     |  0  |    0     |   0   |  0  |  0   |    0.796940851291   |
|    0     |  0  |    0     |   0   |  0  |  0   |    0.984739056527   |
+----------+-----+----------+-------+-----+------+---------------------+
[10 rows x 17 columns]

In [ ]:
diaper_champ_reviews=diaper_champ_reviews.sort('predicted_sentiment',ascending=False)

In [ ]:
diaper_champ_reviews.head()


In [ ]:
selected_words_model.predict(diaper_champ_reviews, output_type='probability')

In [ ]:
selected_words_model.predict(diaper_champ_reviews[0], output_type='probability')

In [ ]:
selected_words_model.predict(diaper_champ_reviews[-1], output_type='probability')

In [ ]:
selected_words_model.predict(diaper_champ_reviews, output_type='probability').tail()

# most positive 0.9984
# most negative 0.32668